### ISSUE TITLE
Docker image published with no digest manifest

### ISSUE THREAD

**Issue by cdloh** [google/ko#415]
I'm having some trouble with ko building a simple go webserver project.

We're using Gitlab Container Registry for stoage and when ko publishes the image it results in the following error from Gitlab once the image has been uploaded

```
Invalid tag: missing manifest digest
```

There doesn't appear to be any issues on the Gitlab side. Was wondering if anybody has seen anything like this in their usage of ko?

**Comment by imjasonh** [comment]
I'm not familiar with gitlab specifically but https://gitlab.com/gitlab-org/gitlab/-/issues/227576 sounds like that error says, basically, "no images are tagged with this tag"

Can you provide the output of the ko command you're running (sensitive names redacted), and any yaml you're specifying, in case that's helpful in understanding the issue?

**Comment by cdloh** [comment]
I've not got any ko configuration just using the base and the command is

```
ko publish . --bare -t <TAG HERE>
```

If I republish the tag it still has a broken manifest.

Full output. (ran from a golang:1.15 docker container)

```
root@bbe22741ee41:/build# ko publish . --bare -t r31421fds
2021/08/11 10:20:22 Using base gcr.io/distroless/static:nonroot for gitlab.work.net/project/path/subpath
2021/08/11 10:20:24 Building gitlab.work.net/project/path/subpath for linux/amd64
2021/08/11 10:20:24 Publishing gitlab.work.net/project/path/subpath:r31421fds
2021/08/11 10:20:26 existing blob: sha256:0629613e65f200efc7f528042b0be72a1633c9fe9a454edc6a841adbe8f28023
2021/08/11 10:20:26 existing blob: sha256:250c06f7c38e52dc77e5c7586c3e40280dc7ff9bb9007c396e06d96736cf8542
2021/08/11 10:20:26 existing blob: sha256:0d7d70899875b098ace120b574f57b39e91415ceaf9b348d6abe168537509f5a
2021/08/11 10:20:26 existing blob: sha256:5db96da604ff6f8899cd8220abc38b1843582ebdee45c60c22dbb068f71f736e
2021/08/11 10:20:26 gitlab.work.net/project/path/subpath:r31421fds: digest: sha256:df489b278cd978af737e34cf4116be1b0e265cd7df2b65795bfd9e67a8d32b65 size: 952
2021/08/11 10:20:26 Published gitlab.work.net/project/path/subpath:r31421fds@sha256:df489b278cd978af737e34cf4116be1b0e265cd7df2b65795bfd9e67a8d32b65
gitlab.work.net/project/path/subpath:r31421fds@sha256:df489b278cd978af737e34cf4116be1b0e265cd7df2b65795bfd9e67a8d32b65
```

Is there a way

**Comment by imjasonh** [comment]
This sounds like some issue related to using `--bare` and `-t` together.

The docs for `-t` hint that this might not be well supported:

```
  -t, --tags strings             Which tags to use for the produced image instead of the default 'latest' tag (may not work properly with --base-import-paths or --bare). (default [latest])
```

...but in my own experience it's never been an issue. I'm able to push to GCR with this for instance:

```
$ KO_DOCKER_REPO=gcr.io/imjasonh/test ko publish ./test --bare -t abcdef
...
gcr.io/imjasonh/test:abcdef@sha256:bc035aae0f9d9f3f1543ad11f97578a0e5cda43af230af608c893349aa8d0136
```

Just to remove variables here, could you try pushing to Gitlab CR without `--bare`? Based on your example that would be something like `KO_DOCKER_REPO=gitlab.work.net/project/path ko publish ./subpath -t r31421fds`. If that works then that narrows down where the issue is.

Another option would be to ko-publish to anywhere, then retag the image to what you want as a separate step. Using [`crane`](https://github.com/google/go-containerregistry/tree/main/cmd/crane):

```
$ crane cp $(ko publish ./) gitlab.work.net/project/path/subpath:r31421fds
```

(If the images are in the same repository, `crane cp` merely retags and doesn't copy/move any blobs)

**Comment by cdloh** [comment]
Okay I did some more testing. (Really appreciate the help here!)

I tested without bare. Still same issue. Tested without a tag (ie just using `latest`) and it also still had the same issue.

I tested by setting the base image to alpine.

my .ko.yaml
```
root@bbe22741ee41:/build# cat .ko.yaml
defaultBaseImage: alpine
root@bbe22741ee41:/build#
```

And this didn't exhibit the issue! The image uploaded successfully, and the tagged image had no digest problems.
